torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -0,0 +1,227 @@
+  // Wait until the send queue is depleted.
+  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  // Use another barrier in case different RpcAgent handles different amounts of
+  // workloads.
+  pg_->barrier()->wait();
+}
+
+std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+    const std::string& to, Message&& message) {
+
+  auto dstRankIter = nameMap_.find(to);
+  TORCH_CHECK(dstRankIter != nameMap_.end(), "Unknown destination worker ", to);
+
+  const int dstRank = dstRankIter -> second;
+  TORCH_CHECK(dstRank != pg_->getRank(), "ProcessGroupAgent does not support "
+    "making RPC calls to self.")
+
+  auto requestId = nextId();
+  auto future = std::make_shared<FutureMessage>();
+  if (message.isRequest()) {
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      futures_[requestId] = future;
+    }
+    message.setId(requestId);
+  } else {
+    future->markCompleted();
+  }
+
+  enqueue(SendWork(dstRank, std::move(message)));
+  return future;
+}
+
+void ProcessGroupAgent::enqueue(SendWork work) {
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  sendQueue_.emplace_back(std::move(work));
+  lock.unlock();
+
+  workProduceCV_.notify_one();
+}
+
+// making sure tensors are not deleted before send finishes
+void ProcessGroupAgent::sendLoop() {
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+
+  while (!stop_) {
+    if (sendQueue_.empty()) {
+      workProduceCV_.wait(lock);
+      continue;
+    }
+
+    auto work = std::move(sendQueue_.front());
+    sendQueue_.pop_front();
+    lock.unlock();
+
+    workConsumeCV_.notify_one();
+
+
+    std::stringstream ss;
+    serialize(work.message_, ss);
+    std::string str = ss.str();
+
+    std::vector<torch::Tensor> preamble = {
+      torch::tensor(
+        {
+          (int64_t)pg_->getRank(),
+          (int64_t)str.length(),
+        }, {torch::kLong})
+    };
+    pg_->send(preamble, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
+    std::vector<torch::Tensor> payload =
+        {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
+    pg_->send(payload, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
+
+    lock.lock();
+  }
+}
+
+void ProcessGroupAgent::listenLoop() {
+  while (true) {
+    // rank, tensor size
+    std::vector<torch::Tensor> preamble = {torch::empty({2}, {torch::kInt64})};
+    pg_->recvAnysource(preamble, pg_->getRank())->wait();
+    int64_t* preamble_items = preamble.front().storage().data<int64_t>();
+
+    auto srcRank = preamble_items[0];
+    auto size = preamble_items[1];
+
+    std::vector<torch::Tensor> tensors = {torch::empty({size}, {torch::kChar})};
+    pg_->recv(tensors, srcRank, pg_->getRank())->wait();
+
+    std::stringstream ss(std::string(
+      (char*)tensors[0].storage().data<signed char>(), tensors[0].numel()));
+
+    Message message = deserialize(ss);
+
+    if (message.isRequest()) {
+      cb_(names_[srcRank], std::move(message), *this);
```

Comment:
@mrshenli 

Call sequence as below.

Let me turn it into a unit test

![image](https://user-images.githubusercontent.com/7608630/62580943-863a9900-b85c-11e9-91de-67af5cf809ae.png)

Reply:
Yep, cb has to be non-blocking for UDFs. It will come in the next PR. For this one, as we only target builtin ops, blocking cb should suffice.